src/audio/midi.rs

```diff
@@ -0,0 +1,146 @@
+use crate::Result;
+use anyhow::anyhow;
+use midir::{MidiInput, MidiInputPort};
+use num_traits::Num;
+use std::sync::mpsc::{self, Receiver};
+use std::thread;
+use std::time::Duration;
+use std::{convert::TryFrom, sync::mpsc::Sender};
+use wmidi::{Channel, ControlFunction, ControlValue, MidiMessage};
+
+/// Listens on the provided port, and sends MIDI messages over a channel. Returns the receiver of that channel.
+pub fn listen_for_input(port_name: &str) -> Result<Receiver<MidiMessage<'static>>> {
+    let port = port(port_name)?;
+
+    println!("MIDI input: {}", port_name);
+
+    let (sender, receiver) = mpsc::channel();
+    handle_messages(port, sender);
+
+    Ok(receiver)
+}
+
+pub fn port_names() -> Result<Vec<String>> {
+    let midi_input = midi_input()?;
+    midi_input
+        .ports()
+        .iter()
+        .map(|port| midi_input.port_name(port).map_err(|e| e.into()))
+        .collect()
+}
+
+/// Get the last value for a `ControlFunction` from a list of messages.
+pub fn latest_control_value(
+    messages: &[MidiMessage<'static>],
+    channel: Channel,
+    control_function: ControlFunction,
+) -> Option<ControlValue> {
+    messages.iter().rev().find_map(|message| match message {
+        MidiMessage::ControlChange(ch, function, value)
+            if ch == &channel && function == &control_function =>
+        {
+            Some(*value)
+        }
+
+        _ => None,
+    })
+}
+
+/// Maps a control value (0-127) into a range.
+/// eg. `control_value_in_range(50, 100, 64) == 75`
+pub fn control_value_in_range<T: Num + From<u8> + Copy>(min: T, max: T, value: ControlValue) -> T {
+    let value: T = from_control_value(value);
+    value * (max - min) / from_control_value(ControlValue::MAX) + min
+}
+
+fn midi_input() -> Result<MidiInput> {
+    Ok(MidiInput::new("Input")?)
+}
+
+fn port(name: &str) -> Result<MidiInputPort> {
+    let names = port_names()?;
+    let midi_input = midi_input()?;
+
+    midi_input
+        .ports()
+        .into_iter()
+        .find(|port| midi_input.port_name(&port) == Ok(name.into()))
+        .ok_or_else(|| {
+            anyhow!(
+                "Could not find a MIDI port with name '{}'. Available ports are:\n{}",
+                name,
+                names.join("\n")
+            )
+        })
+}
+
+fn handle_messages(port: MidiInputPort, sender: Sender<MidiMessage<'static>>) {
+    thread::spawn(move || {
+        // _connection needs to be a named parameter, because it needs to be kept alive until the end of the scope
+        let _connection = midi_input()
+            .unwrap()
+            .connect(
+                &port,
+                "midir-read-input",
+                move |_, message, _| {
+                    if let Some(message) =
+                        MidiMessage::try_from(message).unwrap().drop_unowned_sysex()
+                    {
+                        sender.send(message).unwrap();
```

Comment:
This returns a result, but it'll only fail if the receiver stops existing, which would be a bug, so I'm `unwrap`ping to just crash if it fails.